File: go/interest-is-interesting/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
	if balance < 0 {
		return 3.213
	} else if balance < 1000 {
		return 0.5
	} else if balance < 5000 {
		return 1.621
	} else {
		return 2.475
	}
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
	return balance * float64(InterestRate(balance)) / 100
}

// AnnualBalanceUpdate calculates the annual balance update, taking into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
	return balance + Interest(balance)
}

// YearsBeforeDesiredBalance calculates the minimum number of years required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) (years int) {
	currentBalance := balance
	for {
		if currentBalance >= targetBalance {
			break
		}

		years += 1
		currentBalance = AnnualBalanceUpdate(currentBalance)
	}

	return years
}

File: clojure/interest-is-interesting/src/interest_is_interesting.clj
(ns interest-is-interesting
  "The Interest is Interesting exercise

  https://exercism.org/tracks/clojure/exercises/interest-is-interesting"
  (:require [clojure.math :as math]))

(defn interest-rate
  "Return the interest rate based on the specified balance."
  [balance]
  (condp > balance
    0 -3.213
    1000 0.5
    5000 1.621
    2.475))

(defn annual-balance-update
  "Return the annual balance update, taking into account the interest rate."
  [balance]
  (-> balance
      (* (/ (bigdec (abs (interest-rate balance))) 100M))
      (+ balance)
      bigdec))

(defn amount-to-donate
  "Return the amount to donate based on the balance and tax-free percentage."
  [balance tax-free-percentage]
  (if (pos? balance)
    (-> balance (* 2M (/ tax-free-percentage 100)) math/floor int)
    0))
